### PR TITLE
Sockets - Nidhi

### PR DESCRIPTION
**Pull request by nidhiparixitpatel**
# Hotel
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
:------------- | :-------------
What was a design challenge that you encountered on this project? | There were many design challenges that I encountered. One of the biggest was deciding how many classes to have and what responsibilities they would have. I played around with having block be its own class and then ended up deciding to split block into block reservation and block classes, instead of keeping it as one class. 
What was a design decision you made that changed over time over the project? | At first I had a method in booker that would load the 20 rooms into the rooms list every time an instance of booker was created. I decided later to change that method and put it in room and made it a self.load_rooms so that the creation of rooms was not coupled to the booker class. 
What was a concept you gained clarity on, or a learning that you'd like to share? | I gained clarity on how important and useful design is. It helped so much to think about how I wanted things to relate to one another before just going ahead and coding. 
What is an example of a _nominal test_ that you wrote for this assignment? What makes it a nominal case? | An example of a nominal test is ensuring that a reservation is made on requesting date ranges that are different and that these are added to the reservation list. It is a nominal case because it is an example of how the program should work in a regular situation. 
What is an example of an _edge case test_ that you wrote for this assignment? What makes it an edge case? | I wrote an edge case test for booker_spec where it tests what happens if all the rooms are booked for a given requested date. It is an edge case because it tests what happens at the limits of the code. 
How do you feel you did in writing pseudocode first, then writing the tests and then the code? | Sometimes I would get carried away and want to write code before the tests, but I realized after a few rounds that it is much more efficient to write pseudocode, then tests, and then code because it makes sure that the requirements are met and its easier to spot holes.

### CODE REVIEW

File: lib/block.rb [hunk 1]
```diff
@@ -0,0 +1,36 @@
+module Hotel
+
```
Comment (dHelmgren) [inline lib/block.rb:2]: This is totally not a big deal, but it looks like your tabs are set to be huge! Our in house standard is 2 spaces, you should be able to update this in your VS Code settings.

File: lib/reservation.rb [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+        attr_accessor :room
+
+        def initialize(id:, date_range: , room: nil)
+            @id = id
+            @date_range = date_range
+            @room = room
+            @cost = cost
+        end
+
+        def cost
```
Comment (dHelmgren) [inline lib/reservation.rb:16]: You overrode the what cost meant with this method definition! You no longer have an attr_reader, you just have this method.

File: spec/booker_spec.rb [hunk 3]
```diff
@@ -0,0 +1,241 @@
...
+            expect(booker.view_available_rooms_in_block(new_block).length).must_equal 1
+        end
+
+        it "raises exception when trying to reserve a room in a block and it is already reserved" do
+            new_block = request_block
+            block_reservation1 = booker.reservations[0]
+            booker.reserve_room_in_block(block: new_block, block_reservation: block_reservation1)
+            expect{booker.reserve_room_in_block(block: new_block, block_reservation: block_reservation1)}.must_raise ArgumentError
+        end
+
```
Comment (dHelmgren) [inline spec/booker_spec.rb:232]: You missed at least one important test for the block: Does your program return the correct cost for the reservation in the block?

File: spec/booker_spec.rb [hunk 4]
```diff
@@ -0,0 +1,241 @@
...
+    end
+
+
+    describe "Wave 2 book_room method with multiple rooms" do
+        
+        let(:booker) {Hotel::Booker.new(Hotel::Room.load_rooms)}
+        
+        it "finds different rooms for 5 of the same requesting dates" do
```
Comment (dHelmgren) [inline spec/booker_spec.rb:116]: I like this test a lot!

**Comment by dHelmgren** [comment]
# Hotel
## What We're Looking For


Feature | Feedback
---     | ---
**Baseline** |
Used git regularly | yes
Answer comprehension questions | yes
**Design** |
Each class is responsible for a single piece of the program | yes
Classes are loosely coupled | generally, aside from your booking class.
**Wave 1** |
List rooms | yes
Reserve a room for a given date range | yes
List reservations for a given date | yes
Calculate reservation price | yes
Invalid date range produces an error | yes
Test coverage | Good!
**Wave 2** |
View available rooms for a given date range | yes
Reserving a room that is not available produces an error | yes
Test coverage | Good!
**Wave 3** |
Create a block of rooms | yes
Check if a block has rooms | yes
Reserve a room from a block | yes
Test coverage | Your discount method doesn't work, see comment!
**Fundamentals** |
Names variables, classes and modules appropriately | yes
Understanding of variable scope - local vs instance | yes
Can create complex logical structures utilizing variables | yes
Appropriately uses methods to break down tasks into smaller simpler tasks | yes
Understands the differences between class and instance methods | yes
Appropriately uses iterators and `Enumerable` methods | yes
Appropriately writes and utilizes classes | yes
Appropriately utilizes modules as a namespace | yes
**Wrap Up** |
There is a refactors.txt file | yes
The file provides a roadmap to future changes | yes
**Additional Feedback** | Good work on this, it's a very solid submission. Testing things completely is pretty hard, but keep at it and you'll find that most programs come out better as you test more extensively.
